drivers/display/display_gc9x01x.c

```diff
@@ -0,0 +1,745 @@
+		return -ENOTSUP;
+	}
+
+	retval = gc9x01x_transmit(dev, GC9X01X_CMD_PIXFMT, &tx_data, 1U);
+	if (retval < 0) {
+		return retval;
+	}
+
+	data->pixel_format = pixel_format;
+	data->bytes_per_pixel = bytes_per_pixel;
+
+	return 0;
+}
+
+static int gc9x01x_set_orientation(const struct device *dev,
+				   const enum display_orientation orientation)
+{
+	struct gc9x01x_data *data = dev->data;
+
+	int retval;
+	uint8_t tx_data = GC9X01X_MADCTL_VAL_BGR;
+
+	if (orientation == DISPLAY_ORIENTATION_NORMAL) {
+		/* works 0° */
+		tx_data |= 0;
+	} else if (orientation == DISPLAY_ORIENTATION_ROTATED_90) {
+		/* works CW 90° */
+		tx_data |= GC9X01X_MADCTL_VAL_MV | GC9X01X_MADCTL_VAL_MY;
+	} else if (orientation == DISPLAY_ORIENTATION_ROTATED_180) {
+		/* works CW 180° */
+		tx_data |= GC9X01X_MADCTL_VAL_MY | GC9X01X_MADCTL_VAL_MX | GC9X01X_MADCTL_VAL_MH;
+	} else if (orientation == DISPLAY_ORIENTATION_ROTATED_270) {
+		/* works CW 270° */
+		tx_data |= GC9X01X_MADCTL_VAL_MV | GC9X01X_MADCTL_VAL_MX;
+	}
+
+	retval = gc9x01x_transmit(dev, GC9X01X_CMD_MADCTL, &tx_data, 1U);
+	if (retval < 0) {
+		return retval;
+	}
+
+	data->orientation = orientation;
+
+	return 0;
+}
+
+static int gc9x01x_configure(const struct device *dev)
+{
+	const struct gc9x01x_config *config = dev->config;
+
+	int retval;
+	enum display_pixel_format pixel_format;
+	enum display_orientation orientation;
+
+	/* Set all the required registers. */
+	retval = config->regs_init_fn(dev);
+	if (retval < 0) {
+		return retval;
+	}
+
+	/* pixel format */
+	if (config->pixel_format == GC9X01X_PIXEL_FORMAT_RGB565) {
+		pixel_format = PIXEL_FORMAT_RGB_565;
+	} else {
+		pixel_format = PIXEL_FORMAT_RGB_888;
+	}
+
+	/* Set pixel format. */
+	retval = gc9x01x_set_pixel_format(dev, pixel_format);
+	if (retval < 0) {
+		return retval;
+	}
+
+	/* orientation */
+	if (config->rotation == GC9X01X_ORIENTATION_90) {
+		orientation = DISPLAY_ORIENTATION_ROTATED_90;
+	} else if (config->rotation == GC9X01X_ORIENTATION_180) {
+		orientation = DISPLAY_ORIENTATION_ROTATED_180;
+	} else if (config->rotation == GC9X01X_ORIENTATION_270) {
+		orientation = DISPLAY_ORIENTATION_ROTATED_270;
+	} else {
+		orientation = DISPLAY_ORIENTATION_NORMAL;
+	}
+
+	/* Set display orientation. */
+	retval = gc9x01x_set_orientation(dev, orientation);
+	if (retval < 0) {
+		return retval;
+	}
+
+	if (config->inversion) {
+		/* Display inversion mode. */
+		retval = gc9x01x_transmit(dev, GC9X01X_CMD_INVON, NULL, 0);
+		if (retval < 0) {
+			return retval;
+		}
+	}
+
+	return 0;
+}
+
+static int gc9x01x_set_brightness(const struct device *dev, const uint8_t brightness)
+{
+	const struct gc9x01x_config *config = dev->config;
+
+	if (config->backlight.dev != NULL) {
+		uint32_t step = config->backlight.period / 100U;
+		int retval = pwm_set_pulse_dt(&config->backlight, brightness * step / 2.55);
+
+		if (retval) {
+			LOG_ERR("Failed to set pulse width\n");
+			return retval;
+		}
+		return 0;
+	}
```

Comment:
This is normally implemented only if the display controller itself includes backlight control, so I think this should just return -NOTSUP in this case.

I realize this is needed for the samples to work though so maybe we should have a known node and have the samples be aware of that, regardless, that should be fixed separately from this PR.

Reply:
I am aligned that this is not the right solution, but for the board I am currently working with, backlight is only working when the PWM is initialized. All sample applications would show a dark screen as there is currently no possibility to set an initial PWM in device tree.

IMO a pwm backlight driver would be helpful.(https://github.com/torvalds/linux/blob/9f8413c4a66f2fb776d3dc3c9ed20bf435eb305e/drivers/video/backlight/pwm_bl.c)
Maybe we could make use of the `led_driver_api`?

Reply:
Yeah I think this should use the LED API, how about:
- drop the pwm stuff from here
- define the backlight pwm in the board file as a `pwm-leds` node, for example https://github.com/zephyrproject-rtos/zephyr/pull/67311/files for example:
```
	pwmleds {
		compatible = "pwm-leds";
		backlight_pwm: backlight_pwm {
			pwms = <&ledc0 0 PWM_HZ(250) PWM_POLARITY_NORMAL>;
		};
	};
```
- link the node to a well known chosen, maybe `zephyr,display-backlight`
```

	chosen {
		zephyr,display = &gc9a01;
		zephyr,display-backlight = &backlight_pwm;
	};
```
- set `CONFIG_LED=y` in the LVGL samples
- in the LVGL samples check if the backlight chosen node is defined, if it is find the device with `DEVICE_DT_GET(DT_PARENT(DT_CHOSEN(...)))`, the node offset with `DT_NODE_CHILD_IDX(DT_CHOSEN(...))` and then do a `led_set_brightness(dev, idx, value);` in the application

@faxe1008 @jfischer-no would that make sense?

Linux has a dedicated driver because there's some userspace APIs specifically for backlight but for Zephyr things are a bit simpler, I think we can get away with the existing LED APIs.

Reply:
I have a similar definition in the board layer already, but **I think the missing piece is the initialisation an initial PWM value via devicetree**. It should be possible on board level to set an adequate PWM init value. If it is the application being responsible we would need to rework all existing display related sample applications.

Reply:
yeah I'm suggesting to rework the samples (the last point I made)

Reply:
> Linux has a dedicated driver because there's some userspace APIs specifically for backlight but for Zephyr things are a bit simpler, I think we can get away with the existing LED APIs.
We don't need a full api for that. The `led_driver_api` already features all we need, including `led_set_brightness`.

What do you think about the possibility to extend the `pwm-leds` driver with an init value?

For better clarity it may also a possibility to create more specific variant of the `pwm-leds` driver (e.g. `pwm-backlight`) that features even more functionality related to display backlight control. But I don't think that's needed for now.

Reply:
@faxe1008 @jfischer-no What's your opinion on that:
- Do we prefer to rework the graphics examples to set a default brightness when `zephyr,display-backlight` is specified? Are there additional tasks needed for this solution (e.g.documentation)?
- Or do you think it makes sense to fruther think about a possibility to set a default PWM value in the devicetreee for that boards that only rely on PWM-controller backlight?

Reply:
Since there are cases where the backlight is a device config (settings or similar), having it as a DT default value does not feel like an elegant solution. I get that it makes it more user friendly because there is one less thing to think of. I'd love to be this part of the display API, but for simplicity sake the using the led api is ok imo.

Reply:
As per request I removed the embedded backlight control from the driver ([3c8529d](https://github.com/zephyrproject-rtos/zephyr/pull/67012/commits/3c8529d458c1c3d833a1035cf4963e08dd0facb1)).

However, I would like to point out again that most devices with this display controller have PWM backlighting. This must be set by the application, as there is no option in existing `led_pwm` drivers to set an initial duty_cycle in the devicetree. This means that the existing graphic examples do currently not work by default in this case and the display will stay dark.

I will start working on extending the existing `led_pwm` driver or creating a new backlight driver (compared to linux) to get a working solution for the board I am developing with.

Reply:
> This means that the existing graphic examples do currently not work by default

Yes that needs to be fixed would the proposal I mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/67012#discussion_r1446567202 work? Not sure what I'm asking for is clear but I'm happty to create a PR for that myself if that can help clarifying what I mean.

Reply:
This proposal will not work (at least for my board) as there is no possibility to initialize the PWM with a dutycycle. Basically I am already using the same device tree definition in my board/shield (see #67253).

I already started working on a separate `backlight_pwm` driver that is very similar to `led_pwm`. I am still considering if reusing the led_driver_api is the right way, but I'll keep you posted.

Reply:
What's wrong with initializing the level in the application?

Reply:
For me personally it's mainly inconvenience and I could imagine that this could be a trap-fall for some people.

Reply:
You could say the same for `display_blanking_off()`, if you draw your graphics and forget to unblank you also get a blank screen.

Reply:
Hmm, good point. I will rethink about that. Maybe there are additional advantages in terms of power management having the display driver and some sort of backlight control interoperate.

Reply:
My main concern with auto initializing a backlight at driver init level is that that's certainly going to happen before any application code kicked in and draw anything useful on the screen, and potentially (since pwm and display are independent devices) can happen before the screen has ever been initialized, potentially flashing garbage on the screen during the display initialization sequence. That's quite uncommon for memory based display (like this one) but I see it happening all the time with memoryless displays, the fix for that is to turn on the backlight at the right point in time (which I can't imagine not being in the application code) and I say that it'd be nice to have that in our samples anyway.